### PR TITLE
Implement ephemeral IPs

### PR DESCRIPTION
**Pull request by bnaecker**
- Updates the current external IP allocation query to handle both
  floating and ephemeral IPs, by assuming that the whole port range is
  already reserved for any existing IP address.
- Add public datastore methods for creating SNAT and Ephemeral IPs,
  delegating to private method for the actual query running/handling
- Updates sagas to include UUID generation for external IPs as separate
  steps, for idempotency, and to create Ephemeral IPs if they're
  requested. Also rework instance creation/migration sagas to select the
  Ephemeral IP address, if one was requested, or the SNAT if not.
- Adds optional restriction of IP Pools to a project. This adds the
  project ID or name in a bunch of places, and updates the external IP
  allocation query to only consider pools which are unrestricted, or
  whose project ID matches the one of the instance we're allocating an
  IP for. This relies on a new index on the `instance_external_ip`
  table, which induces an undesirable sorting (by project, not IP), so
  we add a new sorting criterion to the query.
- Adds tests, especially for the external IP table's check constraints
  which verify integrity of the name / description / instance ID for
  different kinds of addresses, and for restriction of an IP pool to a
  project.
- Plumb the external IPs up to Nexus's public API, including instance
  creation and an endpoint for listing external IPs for an instance.
- Adds integration tests for assignment of Ephemeral IPs and authz tests
  for the endpoint(s)

### CODE REVIEW

**Comment by bnaecker** [comment]
Should resolve #1458. There are still some TODOs, which I'll be moving into issues.

File: common/src/sql/dbinit.sql [hunk 1]
```diff
@@ -984,6 +987,15 @@ CREATE TABLE omicron.public.ip_pool_range (
...
     ip_pool_id UUID NOT NULL,
+    /* Optional ID of the project for which this range is reserved.
+     *
+     * NOTE: This denormalizes the tables a bit, since the project_id is
+     * duplicated here and in the parent `ip_pool` table. We're allowing this
+     * for now, since it reduces the complexity of the already-bad IP allocation
+     * query, but we may want to revisit that, and JOIN with the parent table
+     * instead.
+     */
+    project_id UUID,
```
Comment (zephraph) [inline common/src/sql/dbinit.sql:998]: Do we have an issue to capture this rework?

Comment (bnaecker) [reply common/src/sql/dbinit.sql:998]: Which rework are you referring to?

Comment (zephraph) [reply common/src/sql/dbinit.sql:998]: It sounds like that's the real chunk of work to be done, so yes, that. I just worry something like this would be easy to lose track of.

Comment (bnaecker) [reply common/src/sql/dbinit.sql:998]: Will do!

File: nexus/src/external_api/params.rs [hunk 2]
```diff
@@ -315,13 +315,23 @@ pub struct NetworkInterfaceUpdate {
 
 // IP POOLS
 
+// Type used to identify a Project in request bodies, where one may not have
+// the path in the request URL.
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+pub struct ProjectPath {
+    pub organization: Name,
+    pub project: Name,
+}
```
Comment (smklein) [inline nexus/src/external_api/params.rs:324]: This feels a lot like the global / non-global image split. Should we have different endpoints for the "global" vs "non-global" IP pools?

File: smf/sled-agent/config.toml [hunk 3]
```diff
@@ -38,7 +38,8 @@ zpools = [
 # this with whatever value they wish to provide inbound connectivity to guests
 # in their local network, using the current workaround methods in OPTE. See
 # how-to-run.adoc for details on how to determine the value for your network.
-gateway_mac = "00:0d:b9:54:fe:e4"
+# gateway_mac = "00:0d:b9:54:fe:e4"
+gateway_mac = "74:ac:b9:a2:cd:f1"
```
Comment (smklein) [inline smf/sled-agent/config.toml:42]: Is this your local gateway mac? This is minor, but should we revert this change?

**Review by smklein — COMMENTED** [review]
Overall, looks good. Biggest questions relate to undo actions within the sagas.

File: nexus/src/app/sagas/instance_create.rs [hunk 4]
```diff
@@ -141,16 +144,40 @@ fn saga_instance_create() -> SagaTemplate<SagaInstanceCreate> {
...
     template_builder.append(
-        "external_ip",
-        "ExternalIp",
+        "snat_ip_id",
+        "SnatIpId",
         ActionFunc::new_action(
-            sic_allocate_external_ip,
-            sic_allocate_external_ip_undo,
+            sic_allocate_instance_snat_ip_id,
+            sic_allocate_temporary_external_ips_undo,
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/sagas/instance_create.rs:155]: Still working my way through your comments (thanks!) but I think most of them are related to this. The approach here, which I may have botched, was intended to be similar to NICs. We may create multiple records during the IP allocation forward action. In particular, we may successfully create N addresses, and then fail to create address N + 1. But if an action fails, it's undo action is not taken (since it supposedly can only be undone after completing successfully).

That means the action that undoes the (potentially-partial) creation of a bunch of addresses has to be attached to the action _before_ that forward action. So I have an action that creates all the UUIDs, whose undo action is deleting all records with those IDs. The actual creation of the records has no undo action.

James had an alternate strategy here, which also works. Instead of doing multiple fallible operations in one saga node, he adds a fixed set of saga nodes for each possible disk that can be created. Since that's limited by a maximum of 8 disks, there are 8 such nodes, each of which is self-contained and includes the "right" or "expected" undo action. The downside is that there may not be 8 disks, so the saga has some number of actions and undo actions that are no-ops.

Both of these work I think, and we added them around the same time. The solution I arrived at feels more natural to me, but if it's confusing I am happy to move this (and the NIC creation part) to mirror James's approach (in a separate PR).

[the remark this answers is not in the document]
Comment (smklein) [reply nexus/src/app/sagas/instance_create.rs:155]: > Still working my way through your comments (thanks!) but I think most of them are related to this. The approach here, which I may have botched, was intended to be similar to NICs. We may create multiple records during the IP allocation forward action. In particular, we may successfully create N addresses, and then fail to create address N + 1. But if an action fails, it's undo action is not taken (since it supposedly can only be undone after completing successfully).
> 
> That means the action that undoes the (potentially-partial) creation of a bunch of addresses has to be attached to the action _before_ that forward action. So I have an action that creates all the UUIDs, whose undo action is deleting all records with those IDs. The actual creation of the records has no undo action.

Ahhhh okay, this makes sense. I still think it's fairly confusing to read this style in the current saga template format. Having the "undo" action be attached to a (seemingly) unrelated node seems also somewhat risky from a refactoring point-of-view. @davepacheco can chime in if he disagrees, but I *think* the intent was that the "undo action should undo the result of this action if it completes successfully, and each action should be atomic".

For example: `sic_allocate_instance_snat_ip` makes a request to the datastore that should "fully complete, or do nothing". So if we attached the undo action (`sic_allocate_temporary_external_ips_undo`) to the same node performing the operation - the `SnatIp` node below - it'll only be invoked if the atomic allocation completed successfully.

Is `sic_allocate_instance_ephemeral_ips` the problem case here, because it is not atomic?
 
> James had an alternate strategy here, which also works. Instead of doing multiple fallible operations in one saga node, he adds a fixed set of saga nodes for each possible disk that can be created. Since that's limited by a maximum of 8 disks, there are 8 such nodes, each of which is self-contained and includes the "right" or "expected" undo action. The downside is that there may not be 8 disks, so the saga has some number of actions and undo actions that are no-ops.
> 
> Both of these work I think, and we added them around the same time. The solution I arrived at feels more natural to me, but if it's confusing I am happy to move this (and the NIC creation part) to mirror James's approach (in a separate PR).

I *think* the approach taken there is preferable because of the aforementioned action/undo action relationship, but I'm also hoping it gets simplified when subsagas are integrated into steno: https://github.com/oxidecomputer/steno/pull/29

[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/sagas/instance_create.rs:155]: I agree with the points about confusion and refactoring. It does require that the number of nodes be known statically, though. We do here, but that's definitely a choice I'm making for expediency, not because of any intrinsic or policy limitation I'm aware of.

But in any case, I'd prefer to move this to a separate issue tracking that work, and otherwise integrate this PR (unless there's just a bug!). Are you cool with that?

[the remark this answers is not in the document]
Comment (smklein) [reply nexus/src/app/sagas/instance_create.rs:155]: Yeah, I am okay with that (as long as we file an issue to track).

I think I mostly wanted to raise awareness of "there are drawbacks to separating the action from the undo" - and if abiding by that policy is painful today, we should figure out how to make it less painful.

[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/sagas/instance_create.rs:155]: Will make an issue prior to merging. Thanks for flagging it!

[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/sagas/instance_create.rs:155]: https://github.com/oxidecomputer/omicron/issues/1466

File: smf/sled-agent/config.toml (hunk 3 above)
Comment (bnaecker) [reply smf/sled-agent/config.toml:42]: It is indeed, good catch. Thanks.

File: nexus/src/external_api/params.rs (hunk 2 above)
Comment (bnaecker) [reply nexus/src/external_api/params.rs:324]: Maybe? One difference I can see with images is that the IP Pools should still be limited to operators. I don't believe that really impacts the implementation of the authz checks, though. It would just be a different route, and a different way of getting the organization and project names. Even in that case, I believe you can still have a fleet-child object, it's just the route that's scoped. If that sounds right, I'm fine changing it. I don't have a strong preference, assuming only operators can muck with them.

File: nexus/src/db/queries/external_ip.rs [hunk 5]
```diff
@@ -260,27 +285,97 @@ impl NextExternalIp {
...
+        // means we can't use that address for the floating/ephemeral IP).
+        // Something like:
+        //
+        // ON (ip, time_deleted IS NULL) = (candidate_ip, TRUE)
+        //
+        // We'll also need to change how we generate the port sequence subquery.
+        // It should always be 0, basically, but not sure the best way to do
+        // that. SELECT 0? That should work:
+        //
+        // SELECT 0 as candidate_first_port, 65535 as candidate_last_port
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/queries/external_ip.rs:300]: That's a stray comment, missed it since it's just above the block I _do_ want to keep.

File: nexus/src/app/instance.rs [hunk 6]
```diff
@@ -488,11 +492,20 @@ impl super::Nexus {
             .derive_guest_network_interface_info(&opctx, &authz_instance)
             .await?;
 
-        let external_ip = self
+        // Collect the external IPs for the instance.
+        //
+        // For now, we take the Ephemeral IP, if it exists, or the SNAT IP if not.
+        // TODO-correctness: Handle multiple IP addresses
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/instance.rs:498]: Sure thing, done

[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/instance.rs:498]: Also: https://github.com/oxidecomputer/omicron/issues/1467

File: nexus/src/app/instance.rs [hunk 7]
```diff
@@ -488,11 +492,20 @@ impl super::Nexus {
...
+        // TODO-correctness: Handle Floating IPs
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/instance.rs:499]: Yeah, will do. I was planning to switch the partition function used here, to collect all SNAT IPs and then everything else. That's because from OPTE's or Dendrite's point of view, Ephemeral and Floating IPs are the same. But I'll track that in the Floating IP issue, and add that check here.

[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/instance.rs:499]: Actually, that's silly. I'm just going to switch the partition to collect SNAT and everything else _now_.

File: nexus/src/app/sagas/instance_migrate.rs [hunk 8]
```diff
@@ -163,12 +164,23 @@ async fn sim_instance_migrate(
...
         ..old_runtime
     };
-    let external_ip = osagactx
+
+    // Collect the external IPs for the instance.
+    //
+    // For now, we take the Ephemeral IP, if it exists, or the SNAT IP if not.
+    // TODO-correctness: Handle multiple IP addresses
+    // TODO-correctness: Handle Floating IPs
+    let (ephemeral_ips, snat_ip): (Vec<_>, Vec<_>) = osagactx
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/sagas/instance_migrate.rs:173]: Also done as above.

File: nexus/src/db/datastore/instance_external_ip.rs [hunk 9]
```diff
@@ -82,37 +115,42 @@ impl DataStore {
 
     /// Delete all external IP addresses associated with the provided instance
     /// ID.
-    // TODO-correctness: This should be made idempotent.
+    ///
+    /// To support idempotency, such as in saga operations, this method returns
+    /// the number of records deleted, rather than the usual `DeleteResult`.
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/datastore/instance_external_ip.rs:120]: Fair enough, I'll make that clear.

File: nexus/src/app/sagas/instance_create.rs [hunk 10]
```diff
@@ -141,16 +144,40 @@ fn saga_instance_create() -> SagaTemplate<SagaInstanceCreate> {
...
         ),
     );
+    template_builder.append(
+        "snat_ip",
+        "SnatIp",
+        new_action_noop_undo(sic_allocate_instance_snat_ip),
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/sagas/instance_create.rs:161]: Resolved / tracked separately.

File: nexus/src/app/sagas/instance_create.rs [hunk 11]
```diff
@@ -141,16 +144,40 @@ fn saga_instance_create() -> SagaTemplate<SagaInstanceCreate> {
...
+        "EphemeralIpIds",
+        ActionFunc::new_action(
+            sic_allocate_instance_ephemeral_ip_ids,
+            sic_allocate_temporary_external_ips_undo,
+        ),
+    );
+    template_builder.append(
+        "ephemeral_ips",
+        "EphemeralIps",
+        new_action_noop_undo(sic_allocate_instance_ephemeral_ips),
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/sagas/instance_create.rs:179]: Resolved / tracked separately.

File: nexus/src/app/sagas/instance_create.rs [hunk 12]
```diff
@@ -141,16 +144,40 @@ fn saga_instance_create() -> SagaTemplate<SagaInstanceCreate> {
...
+    //
+    // While it's not yet implemented, we need to _attach_ existing Floating
+    // IPs, not create new ones.
+    template_builder.append(
+        "ephemeral_ip_ids",
+        "EphemeralIpIds",
+        ActionFunc::new_action(
+            sic_allocate_instance_ephemeral_ip_ids,
+            sic_allocate_temporary_external_ips_undo,
+        ),
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/app/sagas/instance_create.rs:174]: Resolved / tracked separately.

File: nexus/src/external_api/params.rs (hunk 2 above)
Comment (bnaecker) [reply nexus/src/external_api/params.rs:324]: @rmustacc Any thoughts on this one?

File: common/src/sql/dbinit.sql (hunk 1 above)
Comment (bnaecker) [reply common/src/sql/dbinit.sql:998]: https://github.com/oxidecomputer/omicron/issues/1468

File: nexus/src/external_api/params.rs (hunk 2 above)
Comment (bnaecker) [reply nexus/src/external_api/params.rs:324]: Also, to be clear, I think the project-scoped and top-level versions of this function would be identical. They'd literally call the same function, with an `Option<ProjectPath>` argument contained somewhere.

**Review by smklein — APPROVED** [review]

File: nexus/src/external_api/params.rs (hunk 2 above)
Comment (bnaecker) [reply nexus/src/external_api/params.rs:324]: I talked offline with Robert about this. The short answer is that it may make sense to have a "read-only view" of the route tree under `/ip-pools` also available under a project-scoped route. For example, this would make it easier for users (and probably in the implementation) to find the pools available in the context of a given project.

That said, I've filed https://github.com/oxidecomputer/omicron/issues/1470 to follow-up on it. I'm deferring because this probably depends a lot on how the top-level routes and operator API RFDs (288 and 290, respectively) shake out. We can revisit this then, when we have a better sense of how this will impact the few places this might come up in the API (looking at you, `/images`...).
